Add Payments Thank You page

mindy opened the pull request:
**Changes:**
- Add React Thank You page with feedback form
- Add endpoint, `/api/v1/subscriptions/feedback` that posts GA event 
- Remove Jinja FAQ page
- Update urls to point to React FAQ page instead of Jinja FAQ page
- Move relevant pages to non-wiki site (/payments/, /payments/terms/, /payments/thank-you/)
- Add React Testing Library

**To test:** 
1. Nothing links to this page yet, but you can go direcly to:
http://localhost.org:8000/en-US/payments/thank-you/

2. Type a string less than 5 characters in the Feedback form, should see this error message on submit: 
<img width="632" alt="Screen Shot 2020-03-31 at 2 03 56 AM" src="https://user-images.githubusercontent.com/650/77992363-eacacc80-72f3-11ea-8f57-8d836ac85092.png">

3. If backend request fails, should see a different error message: 
<img width="1131" alt="Screen Shot 2020-03-31 at 1 54 50 AM" src="https://user-images.githubusercontent.com/650/77992208-a2131380-72f3-11ea-9b3e-57a64cc341d6.png">

4. Type a string greater than 5 characters, submit form, should see a Thank you message:
<img width="1086" alt="Screen Shot 2020-03-31 at 1 53 18 AM" src="https://user-images.githubusercontent.com/650/77992177-8c9de980-72f3-11ea-9c47-b38ea0625a79.png">

5. On http://wiki.localhost.org:8000/en-US/profiles/test-super/edit, verify that FAQ link now goes to /payments/:
<img width="793" alt="faq" src="https://user-images.githubusercontent.com/650/77326278-f284e280-6cef-11ea-8575-19f307483a2c.png">

6. On http://wiki.localhost.org:8000/en-US/payments/recurring/management, verify that "Monthly Subscriptions FAQ" links to /payments/#contribute-monthly-faq:
<img width="523" alt="faq2" src="https://user-images.githubusercontent.com/650/77326627-7343de80-6cf0-11ea-9934-538579da5c24.png">

Closes #6581

mindy commented on kuma/static/styles/components/payments/_page.scss:
```diff
@@ -26,15 +34,15 @@
             margin-bottom: 0;
         }
-
-        @media #{$mq-mobile-and-down} {
```
We had a few media queries that were `mobile-and-down` and `desktop-and-up`, and I thought it would be more consistent and clear.to make it mobile first and then add media queries for the larger views. The rearranging in this file is a result of that.

mindy commented on kuma/javascript/src/payments/feedback-form.jsx:
```diff
@@ -0,0 +1,55 @@
+            eventCategory: 'monthly payments',
+            eventAction: feedback,
+            eventLabel: 'thank you page'
+            // eventValue: feedback
```
Have an open question for Kadir about this value, which is why it's commented out at the moment.

mindy replied:
This has been resolved. `eventLabel` will be the feedback from the user.

peterbe commented:
>     1. Nothing links to this page yet, but you can go directly to:
>        http://wiki.localhost.org:8000/en-US/payments/thank-you

I'm guessing this is a typo, right?
The page should be on the read-only site only, right?

As a matter of fact, I believe you can reach that page via both. And I think it shouldn't. It should be 100% on the read-only domain. 

It's crazy enough that the Edit Profile page works on both domains. But just like there are decorators like `@ensure_wiki_domain` I think there should be one for `@ensure_read_only_domain` and we should set that on the subscription landing page (aka. `/payments/`) and the subscription thank you page (aka. `/payments/thank-you`)

peterbe commented:
I think it should be so that http://wiki.localhost.org:8000/en-US/payments/thank-you redirects to http://localhost.org:8000/en-US/payments/thank-you

mindy commented:
@peterbe speaking to Tobin about this task (he was checking in to see if it would make it this sprint, and I said probably not if I have to write this decorator) and perhaps having you write the decorator for `@ensure_read_only_domain`, since you have expertise in this area

escattone commented:
Just FYI, I created the sprint task https://github.com/mdn/kuma/issues/6722 to address the CDN needs of the subscriptions endpoints (including the changes in this PR).

peterbe commented:
Basically, if we intend to make any of the payments pages available on the
read-only site, we need to go ahead with the infra task that @escatone
mentioned above.

Or, we simply keep it all on the Wiki.

I guess “your job” is to figure out which domain and URLs we need and want
and we can worry about Python decorators later.

On Mon, Mar 23, 2020 at 1:58 PM Mindy Tchieu <notifications@github.com>
wrote:

> @peterbe <https://github.com/peterbe> speaking to Tobin about this task
> (he was checking in to see if it would make it this sprint, and I said
> probably not if I have to write this decorator) and perhaps having you
> write the decorator for @ensure_read_only_domain, since you have
> expertise in this area
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/mdn/kuma/pull/6717#issuecomment-602762191>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAGQ4ZZNH6X463E4ZCA6CLRI6PNHANCNFSM4LR5HU4Q>
> .
>
-- 
Peter Bengtsson
Mozilla MDN Web Docs
https://www.peterbe.com

mindy commented on package.json:
```diff
@@ -57,6 +57,7 @@
         "@babel/preset-env": "^7.3.1",
         "@babel/preset-flow": "^7.0.0",
         "@babel/preset-react": "^7.0.0",
+        "@testing-library/react": "^10.0.1",
```
I was trying to avoid adding another dependency and write tests using what we already had (Jest and React Test Renderer), but I find React Testing Library really useful and it's recommended by the React team:

> We recommend using React Testing Library which is designed to enable and encourage writing tests that use your components as the end users do.

peterbe replied:
Yeah, it's a great tool. But I thought that what we were using is the same underlying lib.

mindy replied:
True! `@testing-library/react` creates helper functions for `react-dom/test-utils`. I think it makes tests easier to write and I like that the methods they expose force a certain testing philosophy. When I used `enzyme` (which is still great), by comparison, I felt like I was testing the implementation more than the functionality.

mindy commented:
Pushed a change to keep /payments and /payments/thank-you on the wiki domain. 

Tests are added, so this is ready for a look.

peterbe commented:
> Pushed a change to keep /payments and /payments/thank-you on the wiki domain.

This'll potentially clash with https://github.com/mdn/kuma/pull/6723 won't it because the subscription banner appears on the read-only site and if you click that, it'll take you to the Wiki .

Also, if the payments page is React, why is it on the Wiki? I'm so confused.

peterbe commented on kuma/javascript/src/payments/__tests__/feedback-form.test.jsx:
```diff
@@ -0,0 +1,87 @@
+//@flow
+import React from 'react';
+import { render, fireEvent, cleanup } from '@testing-library/react';
```
Awesome!

mindy replied:
Thank you! I like how this test turned out and makes me appreciate `testing-library/react` even more

peterbe commented on kuma/javascript/src/payments/feedback-form.jsx:
```diff
@@ -0,0 +1,56 @@
+                name="feedback"
+                value={feedback}
+                onChange={handleChange}
+                required
```
Not very optional then, is it? :)

schalkneethling replied:
I notice that you can submit and empty string as valid feedback. Or essentially anything as long as there is at least one character.

mindy replied:
Good catch-- are there guidelines around what we should accept, e.g. string length?

peterbe replied:
I think a good rule of them is that it needs to be something more than `feedback.trim()`. So an empty string or a bunch of whitespace should NOT submit.

schalkneethling replied:
Agreed not being able to submit empty feedback is a good start. The next level up is always tricky. We do want useful feedback I reckon so maybe a 10 character minimum? Not sure who should make that call though :( I would say trust your gut on this one

mindy replied:
Thank you both-- I added a rule of minimum 10 characters.

mindy commented:
> > Pushed a change to keep /payments and /payments/thank-you on the wiki domain.
> 
> This'll potentially clash with #6723 won't it because the subscription banner appears on the read-only site and if you click that, it'll take you to the Wiki .
> 
> Also, if the payments page is React, why is it on the Wiki? I'm so confused.

I'm confused too. I moved the pages back to the Wiki because I thought the problem was that we're going back-and-forth between read-only and Wiki. I spoke with Ryan and Tobin and we are going to discuss this tomorrow.

schalkneethling commented:
@mindy Thank you for all this great work!

Currently, one Dockerized test is failing. See below:


```
================================== FAILURES ===================================

_____________________________ test_payments_index ______________________________

Traceback (most recent call last):

  File "/app/kuma/payments/tests/test_views.py", line 29, in test_payments_index

    assert response.status_code == 200

AssertionError: assert 301 == 200

 +  where 301 = <HttpResponsePermanentRedirect status_code=301, "text/html; charset=utf-8", url="http://wiki.localhost/en-US/payments/">.status_code

================== 1 failed, 1982 passed in 201.00s (0:03:21) ==================
```

peterbe commented:
> @mindy Thank you for all this great work!
> 
> Currently, one Dockerized test is failing. See below:
> 
> ```
> ================================== FAILURES ===================================
> 
> _____________________________ test_payments_index ______________________________
> 
> Traceback (most recent call last):
> 
>   File "/app/kuma/payments/tests/test_views.py", line 29, in test_payments_index
> 
>     assert response.status_code == 200
> 
> AssertionError: assert 301 == 200
> 
>  +  where 301 = <HttpResponsePermanentRedirect status_code=301, "text/html; charset=utf-8", url="http://wiki.localhost/en-US/payments/">.status_code
> 
> ================== 1 failed, 1982 passed in 201.00s (0:03:21) ==================
> ```

That's, as you can guess, is hinged on deciding where the subscription stuff should go. Wiki or the read-only site.

peterbe commented:
One important thing to consider is that if you submit the feedback via Google Analytics client-side, only the percentage of people who Google Analytics will have their feedback sent. 
What we could easily do is make the payments thank you page in Django accept a POST request which is the feedback. And on that Django view POST, we can send the feedback to Google Analytics using Python (`ga_tracking.py`) so we'll get 100% of the feedback sent to us.

peterbe commented on kuma/javascript/src/payments/feedback-form.jsx:
```diff
@@ -0,0 +1,53 @@
+                value={feedback}
+                onChange={handleChange}
+                required
+                disabled={!!submitted}
```
It's already a boolean
```suggestion
                disabled={submitted}
```

schalkneethling commented on kuma/javascript/src/payments/feedback-form.jsx:
```diff
@@ -0,0 +1,95 @@
+    //     eventCategory: 'monthly payments',
+    //     eventAction: 'feedback',
+    //     eventLabel: feedback.trim()
+    // });
```
I guess we can safely remove this?

mindy commented:
This is ready for another look! Updated the front-end tests and added a back-end test. I know it's the weekend, but just wanted to get this queued up for Monday :)

mindy commented on package.json:
```diff
@@ -48,15 +48,16 @@
-        "react": "^16.8.4",
-        "react-dom": "^16.8.4"
+        "react": "^16.13.1",
+        "react-dom": "^16.13.1"
```
I kept running into warnings when I ran my Jest tests and the Internet said that upgrading to the latest would fix it, so this and the Jest version bump a few lines below are a result of that.

peterbe replied:
This means we now have to test more carefully on everything since, sadly, this is a pretty severe change. 
We just need to make a mental note of testing everything more careful before merging this PR.

mindy replied:
I know I don't have as much historical context, but I don't see this as a severe change, since they are minor updates, which are inherently assumed to be additive and non-breaking. From a quick glance at the React change log between 16.8.4 and latest, I don't see any glaring issues, but happy to test the rest of the React site more thoroughly.

peterbe commented:
I personally think 10 characters is too much to require. "Thanks!!" is 8 characters. 
My vote is that we change to min 1 non-whitespace characters.

peterbe commented:
There is something busted with these changes to `package.json` and `package-lock.json`. When I go into a bash shell on this branch and from there run `npm install` it triggers an [edit to the `package-lock.json`](https://gist.github.com/peterbe/91211a6394f0934ba9bff9dd2774d632). 
How did you do those changes to `package.json`? I suspect you might have used the `npm` executable of your host rather than the `npm` executable inside the docker image. 

I can tell because different OSes trigger different URLs in the `package-lock.json`. But we should all be using the OS that is in the Docker image. Yes, we should probably switch to `yarn` but that's for another day.

peterbe commented on kuma/javascript/src/payments/feedback-form.jsx:
```diff
@@ -0,0 +1,106 @@
+                            <a
+                                target="_blank"
+                                rel="noopener noreferrer"
+                                href="mailto:mdn-support@mozilla.com"
```
@schalkneethling What's the latest in terms of getting the value of `settings/common.py`'s `CONTRIBUTION_SUPPORT_EMAIL` into the React code?
Honestly, we can wait with this we work on https://github.com/mdn/kuma/issues/6654 but would you mind @mindy putting in a code comment here that mentions something like:
```
/* TODO: not hardcode the email here in code. See https://github.com/mdn/kuma/issues/6654 */
```

mindy replied:
Added the comment!

schalkneethling replied:
The value lives here:
https://github.com/mdn/kuma/blob/cf6743f5d903e0e0c67f0a9ebcb2a489667a533f/kuma/settings/common.py#L1739

To get at that in React, add the following to `jinja2/includes/config.html`
Perhaps here: https://github.com/mdn/kuma/blob/cf6743f5d903e0e0c67f0a9ebcb2a489667a533f/jinja2/includes/config.html#L26

```
win.mdn.contributionSupportEmail = '{{ settings. CONTRIBUTION_SUPPORT_EMAIL }}';
```

you can then access it here as:

```
window.mdn.contributionSupportEmail;
```

Of course, this will not work in the SSR world but, this code will be reached on a form submit so, we will be in browser land.

mindy replied:
Thank you for these instructions!

peterbe commented on kuma/javascript/src/payments/feedback-form.jsx:
```diff
@@ -0,0 +1,106 @@
+                        )}
+                        emailLink={
+                            <a
+                                target="_blank"
```
Do you need to do that on a `mailto:` hyperlink??

mindy replied:
I think so! I use the browser version of Gmail, so when I remove the `_blank`, it opens up Gmail in the same tab, versus opening up Gmail in a new tab.

peterbe commented on kuma/javascript/src/payments/feedback-form.jsx:
```diff
@@ -0,0 +1,106 @@
+    return (
+        <form onSubmit={handleSubmit}>
+            <input
+                data-testid="feedback-input"
```
Nice!

mindy replied:
Thank you!

peterbe commented on kuma/javascript/src/payments/feedback-form.jsx:
```diff
@@ -0,0 +1,106 @@
+            <input
+                data-testid="feedback-input"
+                type="text"
+                placeholder={gettext('Enter optional feedback…')}
```
This still doesn't make sense. Or am I missing something? The `<input required>` means the browser will attempt to block the `onsubmit` if the `input` is left empty.

mindy replied:
It's optional to submit feedback, so the user can choose ignore the form if they want to. I think it makes more sense in the context of the whole page: 

<img width="1121" alt="Screen Shot 2020-03-30 at 10 19 46 PM" src="https://user-images.githubusercontent.com/650/77980154-97955180-72d4-11ea-87ce-83940e51545b.png">

Does that help?

peterbe commented on kuma/payments/urls.py:
```diff
@@ -3,13 +3,8 @@
-        name="recurring_payment_subscription",
-    ),
     re_path(r"^terms/?$", views.payment_terms, name="payment_terms"),
+    re_path(r"^thank-you/?$", views.thank_you, name="thank_you"),
```
I think we should be consistent about the trailing slashes. I'm perfectly fine with ALWAYS having a trailing `/` by default. The only reason to use `/?` is if you have some historical legacy.

mindy replied:
Good point-- I changed the urls to always have a trailing `/`

peterbe commented on kuma/payments/views.py:
```diff
@@ -23,21 +31,33 @@ def index(request):
+    backend to ensure that all feedback is collected, even
+    from users with DNT or where GA is disabled.
+    """
+    data = json.loads(request.body)
```
We should be consistent and do a check here too. At least something like this:
```python
feedback = (data.get('feedback') or '').strip()
if not feedback:
    return HttpResponseBadRequest("no 'feedback'")
```

mindy replied:
Added, thank you!

peterbe commented on kuma/urls.py:
```diff
@@ -147,6 +148,7 @@
     ),
     # We use our own views for setting language in cookies. But to just align with django, set it like this.
     re_path(r"^i18n/setlang/", core_views.set_language, name="set-language-cookie"),
+    re_path(r"^payments/feedback/?$", send_feedback, name="send_feedback"),
```
Ditto.

mindy replied:
Removed the `?`

peterbe commented:
To untangle the "corruption" of the `package-lock.json` I think you need to start over a bit on that. The only way I know is that you reset `package.json` AND `package-lock.json` from `master`.  Write down which upgrades and additions you want "on a piece of paper" and the next time, use `docker-compose exec web bash`. Actually, I think I can help.

mindy commented on kuma/javascript/src/payments/__tests__/thank-you.test.jsx:
```diff
@@ -0,0 +1,13 @@
+//@flow
```
I agree with you in that it feels redundant. I prefer having things in a separate `__test__` folder because I think it's more organized, especially when there's snapshots. And I noticed all the tests had `.test.` like you mentioned, so this was an attempt to have both consistency with the existing test file names and hide the clutter behind a folder. 

Any way, I moved them to /payments and got rid of the `__test__` folder.

mindy commented on kuma/javascript/src/payments/__tests__/thank-you.test.jsx:
```diff
@@ -0,0 +1,13 @@
+
+describe('Payments Thank You page', () => {
+    test('it renders', () => {
+        const renderer = new ShallowRenderer();
```
I originally used it because I care mainly about whether or not the high-level items rendered, like `<FeedbackForm />`. 

I don't find snapshot testing very useful, so getting rid of the `__tests__` folder (since snapshots don't have a place to hide) encouraged me to re-write this test so it doesn't rely on snapshots.

mindy commented on kuma/javascript/src/payments/feedback-form.jsx:
```diff
@@ -0,0 +1,106 @@
+
+                if (!res.ok) {
+                    throw new Error(
+                        'Request (POST) to /api/v1/payments/feedback failed'
```
Fixed-- thanks for catching! I kept going back and forth and wasn't sure where this url should go.

schalkneethling commented on kuma/static/styles/components/payments/_page.scss:
```diff
@@ -73,82 +88,119 @@
+                button {
+                    // this keeps button flush right
+                    margin-left: auto;
+                }
```
Nit: There should be one empty line between each of these blocks.

mindy replied:
Fixed, thank you!

schalkneethling commented:
@mindy following error in Travis:

```
File "/app/kuma/urls.py", line 17, in <module>
    from kuma.payments.urls import lang_urlpatterns as payments_lang_urlpatterns
  File "/app/kuma/payments/urls.py", line 3, in <module>
    from . import views
  File "/app/kuma/payments/views.py", line 13, in <module>
    from kuma.core.ga_tracking import (
ImportError: cannot import name 'ACTION_FEEDBACK' from 'kuma.core.ga_tracking' (/app/kuma/core/ga_tracking.py)
```

mindy commented:
> @mindy following error in Travis:
> 
> ```
> File "/app/kuma/urls.py", line 17, in <module>
>     from kuma.payments.urls import lang_urlpatterns as payments_lang_urlpatterns
>   File "/app/kuma/payments/urls.py", line 3, in <module>
>     from . import views
>   File "/app/kuma/payments/views.py", line 13, in <module>
>     from kuma.core.ga_tracking import (
> ImportError: cannot import name 'ACTION_FEEDBACK' from 'kuma.core.ga_tracking' (/app/kuma/core/ga_tracking.py)
> ```
> 
> Should it perhaps be `ACTION_SUBSCRIPTION_FEEDBACK`

Thanks Schalk! Just pushed lots of commits to address this issue, as well as the other comments.

mindy commented on kuma/javascript/src/__mocks__/mockSvg.js:
```diff
@@ -1 +1 @@
-export default 'svg mock';
+export default 'svg-mock';
```
`react-testing-library` throws an "Invalid syntax" error when rendering this, so this fixes it. It doesn't like the space, and the Jest docs also have a hyphenated string.

mindy commented:
> Ideally, I think the feedback submission so be done in `/api/v1/subscription/feedback/` instead of the existing Django view. But I understand the compromise. If you need help, reach out to me or @Gregoor as [his ongoing PR](https://github.com/mdn/kuma/pull/6756) of uses the `api/v1` stuff.
> 
> One thing that is a bit weird is that the input field becomes disabled after you've submitted a piece of feedback. But the "Send" button isn't disabled. That's inconsistent. You're allowed to press the button a second time and then this happens:
> <img alt="Screen Shot 2020-03-30 at 4 12 21 PM" width="724" src="https://user-images.githubusercontent.com/26739/77957537-a282be80-72a1-11ea-8f9b-b412bfbf3b91.png">
> 
> If we're confident we don't want to allow multiple submissions we should _not_ render the form after a successful submission, and just show the "Thank you for submitting your feedback!" text entirely in lieu of the form.

Getting lost in all of the comments and just noticed this. 
- regarding /api/v1: I originally placed it there, as you can see from the comment, but kept going back and forth about it and wasn't sure what pattern to follow. I can make this change tomorrow. 
- regarding disabiling the text field: I agree that it's a bit weird-- I originally had it enabled, and it would clear the "Thank you" message if you started typing in the field again, but changed it to meet the acceptance criteria. I left a comment for Kadir in the ticket, asking why we want to disable the form.
